main.py

```diff
@@ -700,6 +710,28 @@ def delete_schedule(self, request: Request) -> JSONResponse:
         log.debug("delete_schedule result %s %s", result, status)
         return JSONResponse(result, status_code=status)
 
+    def _check_no_tag_duplication(self, evc_dict: dict):
+        """Check if the given EVC has UNIs with no tag and if these are
+         duplicated. Raise DuplicatedNoTagUNI if duplication is found.
+        Args:
+            evc (dict): EVC to be analyzed.
+        """
+        uni_a, uni_z = evc_dict.get("uni_a"), evc_dict.get("uni_z")
+
+        # No UNIs
+        if not (uni_a or uni_z):
+            return
+
+        if (not (uni_a and not uni_a.user_tag) and
+                not (uni_z and not uni_z.user_tag)):
+            return
+        for circuit in self.circuits.values():
```

Comment:
To be safe since this is being called in a threaded context, `self.circuits` dict size might change so you'll need a shallow copy to avoid a potential exception. Which also implies that there's room for a potential race condition bug too (conflicting EVCs being created concurrently), but the old implementation also had this other potential concurrency creation issue too so it's not introducing a new bug, which is also very rare and controlled case if it were to ever happen, so a shallow copy for now should suffice, and then in the future we can also further isolate with a new lock protecting `self.circuits` before iterating and changing it, see what I mean? So, let's only go for the shallow copy for now.

Reply:
Gotta, added copy.